src/requested_path.rs

```diff
@@ -13,15 +14,27 @@ fn decode_percents(string: &&str) -> String {
     percent_decode(string.as_bytes()).decode_utf8().unwrap().into_owned()
 }
 
+fn normalize_path(path: &Path) -> PathBuf {
+    path.components().fold(PathBuf::new(), |mut result, p| {
+        match p {
+            Component::Normal(x) => {
+                result.push(x);
+                result
+            }
+            Component::ParentDir => {
+                result.pop();
+                result
+            },
+            _ => result
+        }
+    })
+}
+
 impl RequestedPath {
     pub fn new<P: AsRef<Path>>(root_path: P, request: &Request) -> RequestedPath {
+        let decoded_req_path = PathBuf::from_iter(request.url.path().iter().map(decode_percents));
```

Comment:
If we don't have permission for a file it is nonexistent for our purposes IMO.

(BTW the behavior for "Permission Denied" is wrong IMO, bad file permissions rather implies a 500 Internal Server Error. But 404 is fine too I think)